src/geant4/app/src/ExN01DetectorConstruction.cc

```diff
@@ -0,0 +1,454 @@
+          filter->add(pyne::particle::geant4(scorer.particle_name));
+          particle_filters[scorer.particle_name]=filter;
+        }
+      */
+    }
+  }
+
+  // prints the maps
+  std::map<int, std::vector<std::string> >::iterator it;
+  for (it = volume_part_map.begin() ; it != volume_part_map.end() ; ++it) {
+    std::vector<std::string>::iterator str;
+    for (str = (it->second).begin() ; str != (it->second).end() ; ++str) {
+      std::cout << (it->first) << " " << (*str) << std::endl;
+    }
+  }
+
+  // build the filters
+
+  // setup the data for the detectors histograms
+  build_histogram();
+
+
+  int sd_index = 0; // the number of sensitive detectors
+  //  loop over the volume indices
+  for (it = volume_part_map.begin() ; it != volume_part_map.end() ; ++it) {
+    EntityHandle vol = dagmc->entity_by_id(3, it->first); // convert id to eh
+    int vol_idx = dagmc->index_by_handle(vol); // get the volume index
+
+    // turn the idx into string
+    std::stringstream int_to_string;
+    int_to_string << vol_idx;
+    std::string idx_str = int_to_string.str();
+
+    // get detector name
+    std::string detector_name = "vol_" + idx_str + "_flux";
+
+    // increment the detector counte
+    ++sd_index;
+
+    // loop over the vector of particle types
+    std::vector<std::string>::iterator str;
+    std::vector<std::string> particle_types = (it->second);
+    for (str = particle_types.begin() ; str != particle_types.end() ; ++str) {
+      // particle name
+      std::string particle_name = *str;
+
+      // create new detector
+      G4cout << detector_name + '_' + particle_name << G4endl;
+
+      int pdc = pyne::particle::id(particle_name);
+      // if pdc = 0, is heavy ion, set pdc as nucid
+      if (pdc == 0)
+        pdc = pyne::nucname::id(particle_name);
+
+      // create a histogram for each particle
+      add_histogram_description(detector_name + "_" + particle_name);
+      HM->add_histogram(sd_index, pdc);
+    }
+
+    // get detector volume
+    double det_volume = dag_logical_volumes[vol_idx]->GetSolid()->GetCubicVolume();
+
+    // create new detector
+    G4VSensitiveDetector* detector = new ExN01SensitiveDetector(detector_name,
+                                                                "flux", sd_index, det_volume * cm3,
+                                                                HM);
+
+    // May wish to filter particles here as well
+
+    //sets the sensitivity
+    // build the filters
+    /*
+    for ( str = particle_types.begin() ; str != particle_types.end() ; ++str )
+      {
+        // particle name
+        std::string particle_name = *str;
+        detector->SetFilter(particle_filters[particle_name]);
+      }
+     */
+    // add the detector
+    G4SDManager::GetSDMpointer()->AddNewDetector(detector);
+    dag_logical_volumes[vol_idx]->SetSensitiveDetector(detector);
+  }
+  end_histogram();
+
+  HM->print_histogram_collection();
```

Comment:
Memory Leak: `HM` should be `delete`'d, here and for any other return values for this method.